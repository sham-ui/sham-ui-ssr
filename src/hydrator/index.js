import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,

        // Pass default hooks service for proxy some methods (rehydrate & resolveID)
        new ServerSideHooks( DI.resolve( HOOKS_DI_KEY ), storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,

        // Don't pass default dom service, because ServerSideDOM full override it
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
