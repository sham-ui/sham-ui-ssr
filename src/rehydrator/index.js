import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
