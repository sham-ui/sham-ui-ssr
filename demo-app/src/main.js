import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

