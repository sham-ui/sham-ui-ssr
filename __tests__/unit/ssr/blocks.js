import { compile, compileAsSFC } from 'sham-ui-test-helpers';
import { ssr } from '../../../src/testing';

beforeEach( () => {
    window.LinkTo = compile`
        <a href={{url}}>
            {% defblock %}
        </a>
    `;

    window.DisplayContent = compile`
        {% if condition %}
            {% defblock %}
        {% endif %}
    `;
    window.TextContent = compile`
        <span>
            {% defblock %}
        </span>
    `;
} );

afterEach( () => {
    delete window.LinkTo;
    delete window.DisplayContent;
    delete window.TextContent;
} );

it( 'should work with {% block "default" %}', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <div>
                <LinkTo>
                    {% default %}
                        Text for content
                    {% end default %}
                </LinkTo>
            </div>
        `,
        {}
    );
    expect( meta.html ).toBe(
        '<div><a> Text for content <!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with two named blocks', async() => {
    expect.assertions( 2 );
    window.CustomPanel = compile`
        <div>
            <div class="title">
                {% defblock title %}
            </div>
            <div class="content">
                {% defblock %}
            </div>
        </div>
    `;
    const meta = await ssr(
        compile`
            <div>
                <CustomPanel>
                    {% title %}
                        Text for title
                    {% end title %}

                    {% default %}
                        Text for content
                    {% end default %}
                </CustomPanel>
            </div>
        `,
        {}
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div><div><div class="title"> Text for title <!--0--></div><div class="content"> Text for content <!--1--></div></div></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.CustomPanel;
} );

it( 'should work with component arguments', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <div>
                <LinkTo url={{url}}>
                    {% default %}
                        Text for {{url}}
                    {% end default %}
                </LinkTo>
            </div>
        `,
        {
            url: 'http://example.com'
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div><a href="http://example.com"> Text for <!--0-->http://example.com<!--1--> <!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with component default block', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <div>
                <LinkTo url={{url}}>
                    Text for {{url}}
                </LinkTo>
            </div>
        `,
        {
            url: 'http://example.com'
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div><a href="http://example.com"> Text for <!--0-->http://example.com<!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should remove block in if', async() => {
    expect.assertions( 2 );
    window.VisibleBlock = compile`
        {% if visible %}
            <div class="content">
                {% defblock %}
            </div>
        {% endif %}
    `;
    const meta = await ssr(
        compile`
            <VisibleBlock visible={{visible}}>
                Text content for {{data}}
            </VisibleBlock>
        `,
        {
            visible: true,
            data: 'foo'
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div class="content"> Text content for <!--0-->foo<!--0--></div><!--0--><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.VisibleBlock;
} );

it( 'should work with two nested if', async() => {
    expect.assertions( 2 );
    window.BigRedButton = compile`
        {% if big %}
            {% if red %}
                <button class="big red">This button big={{big}}, red={{red}}{% defblock %}</button>
            {% endif %}
        {% endif %}
    `;
    const meta = await ssr(
        compile`
            <BigRedButton big={{big}} red={{red}}>
                big && red
            </BigRedButton>
        `,
        {
            big: false,
            red: false
        }
    );
    expect( meta.html ).toBe( '<!--0--><!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.BigRedButton;
} );

it( 'should work with defblock nested in useblock', async() => {
    expect.assertions( 2 );
    window.LoadedContainer = compile`
        {% if loaded %}
            {% defblock %}
        {% endif %}
    `;
    window.LoadedVisibleContainer = compile`
        <LoadedContainer loaded={{loaded}}>
            {% if visible %}
                {% defblock %}
            {% endif %}
        </LoadedContainer>
    `;
    window.RedLoadedVisibleContainer = compile`
        <LoadedVisibleContainer loaded={{loaded}} visible={{visible}}>
            {% if red %}
                {% defblock %}
            {% endif %}
        </LoadedVisibleContainer>
    `;
    const meta = await ssr(
        compile`
            <RedLoadedVisibleContainer loaded={{loaded}} visible={{visible}} red={{red}}>
                red && loaded & visible
            </RedLoadedVisibleContainer>
        `,
        {
            red: false,
            loaded: false,
            visible: false
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<!--0--><!--0--><!--0--><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.LoadedContainer;
    delete window.LoadedVisibleContainer;
    delete window.RedLoadedVisibleContainer;
} );

it( 'should work with for', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            <ul>
                {% for url of links %}
                    <li>
                        <LinkTo url={{url}}>Text for {{url}}</LinkTo>
                    </li>
                {% endfor %}
            </ul>
            <LinkTo url="http://example.com">Home</LinkTo>
        `,
        {
            links: [
                'http://foo.example.com',
                'http://bar.example.com',
                'http://baz.example.com'
            ]
        }
    );
    expect( meta.html ).toBe(
        '<ul>' +
            // eslint-disable-next-line max-len
            '<li><a href="http://foo.example.com">Text for <!--0-->http://foo.example.com<!--0--></a></li>' +
            // eslint-disable-next-line max-len
            '<li><a href="http://bar.example.com">Text for <!--0-->http://bar.example.com<!--0--></a></li>' +
            // eslint-disable-next-line max-len
            '<li><a href="http://baz.example.com">Text for <!--0-->http://baz.example.com<!--0--></a></li>' +
        '</ul>' +
        '<a href="http://example.com">Home<!--0--></a><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work useblock if was update from block component', async() => {
    expect.assertions( 1 );
    const meta = await ssr(
        compile`
            <DisplayContent>
                Content
            </DisplayContent>
        `,
        {

        }
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should correct resolve owner', async() => {
    expect.assertions( 1 );
    const meta = await ssr(
        compileAsSFC`
            <template>
                <TextContent>
                    <TextContent>
                        {{this._text()}}
                    </TextContent>
                </TextContent>
            </template>

            <script>
                export default Component( Template, function( options ) {
                    const state = options( {
                        text() {
                            return 'Text for content'    
                        }
                    } );
                    this._text = () => state.text(); 
                } );
            </script>
        `
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );


