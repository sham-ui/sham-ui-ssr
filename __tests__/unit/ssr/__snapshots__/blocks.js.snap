// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correct resolve owner 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                2
            ]
        ]
    },
    \\"2\\": {
        \\"C\\": [
            \\"3\\"
        ],
        \\"N\\": [
            1
        ]
    },
    \\"3\\": {
        \\"C\\": [
            \\"4\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                1
            ]
        ]
    },
    \\"4\\": {
        \\"N\\": [
            0
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            1
        ]
    }
}",
  "html": "
<span><span>Text for content<!--0--></span>
<!--0-->
<!--0--></span>
<!--0-->
",
}
`;

exports[`should remove block in if 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"O\\": {
            \\"visible\\": true
        },
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            1
        ]
    },
    \\"2\\": {
        \\"C\\": [
            \\"3\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                3
            ]
        ]
    },
    \\"3\\": {
        \\"N\\": [
            0,
            2
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            2
        ]
    }
}",
  "html": "
<div class=\\"content\\"> Text content for
  <!--0-->foo
  <!--0-->
</div>
<!--0-->
<!--0-->
",
}
`;

exports[`should work useblock if was update from block component 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"N\\": [
            0
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            1
        ]
    }
}",
  "html": "
<!--0-->
<!--0-->
",
}
`;

exports[`should work with {% block "default" %} 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                1
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            0
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>
  <a> Text for content
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with component arguments 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"O\\": {
            \\"url\\": \\"http://example.com\\"
        },
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                5
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            0,
            2,
            4
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>
  <a href=\\"http://example.com\\"> Text for
    <!--0-->http://example.com
    <!--1-->
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with component default block 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"O\\": {
            \\"url\\": \\"http://example.com\\"
        },
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                3
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            0,
            2
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>
  <a href=\\"http://example.com\\"> Text for
    <!--0-->http://example.com
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with defblock nested in useblock 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"O\\": {
            \\"loaded\\": false,
            \\"visible\\": false,
            \\"red\\": false
        },
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            2
        ]
    },
    \\"2\\": {
        \\"O\\": {
            \\"loaded\\": false,
            \\"visible\\": false
        },
        \\"C\\": [
            \\"3\\"
        ],
        \\"N\\": [
            1
        ]
    },
    \\"3\\": {
        \\"O\\": {
            \\"loaded\\": false
        },
        \\"N\\": [
            0
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            3
        ]
    }
}",
  "html": "
<!--0-->
<!--0-->
<!--0-->
<!--0-->
",
}
`;

exports[`should work with for 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"O\\": {
            \\"url\\": \\"http://example.com\\"
        },
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            1,
            [
                1,
                1
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            0
        ]
    },
    \\"3\\": {
        \\"C\\": [
            \\"6\\"
        ],
        \\"N\\": [
            0
        ]
    },
    \\"4\\": {
        \\"C\\": [
            \\"8\\"
        ],
        \\"N\\": [
            1
        ]
    },
    \\"5\\": {
        \\"C\\": [
            \\"10\\"
        ],
        \\"N\\": [
            2
        ]
    },
    \\"6\\": {
        \\"O\\": {
            \\"url\\": \\"http://foo.example.com\\"
        },
        \\"C\\": [
            \\"7\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                3
            ]
        ]
    },
    \\"7\\": {
        \\"N\\": [
            0,
            2
        ]
    },
    \\"8\\": {
        \\"O\\": {
            \\"url\\": \\"http://bar.example.com\\"
        },
        \\"C\\": [
            \\"9\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                3
            ]
        ]
    },
    \\"9\\": {
        \\"N\\": [
            0,
            2
        ]
    },
    \\"10\\": {
        \\"O\\": {
            \\"url\\": \\"http://baz.example.com\\"
        },
        \\"C\\": [
            \\"11\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                3
            ]
        ]
    },
    \\"11\\": {
        \\"N\\": [
            0,
            2
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\",
            \\"3\\",
            \\"4\\",
            \\"5\\"
        ],
        \\"N\\": [
            0,
            2
        ]
    }
}",
  "html": "
<ul>
  <li>
    <a href=\\"http://foo.example.com\\">Text for
      <!--0-->http://foo.example.com
      <!--0-->
    </a>
  </li>
  <li>
    <a href=\\"http://bar.example.com\\">Text for
      <!--0-->http://bar.example.com
      <!--0-->
    </a>
  </li>
  <li>
    <a href=\\"http://baz.example.com\\">Text for
      <!--0-->http://baz.example.com
      <!--0-->
    </a>
  </li>
</ul>
<a href=\\"http://example.com\\">Home
  <!--0-->
</a>
<!--0-->
",
}
`;

exports[`should work with two named blocks 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"C\\": [
            \\"2\\",
            \\"3\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                0
            ],
            [
                0,
                0,
                1
            ],
            [
                0,
                1
            ],
            [
                0,
                1,
                1
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            0
        ]
    },
    \\"3\\": {
        \\"N\\": [
            0
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<div>
  <div>
    <div class=\\"title\\"> Text for title
      <!--0-->
    </div>
    <div class=\\"content\\"> Text for content
      <!--1-->
    </div>
  </div>
</div>
",
}
`;

exports[`should work with two nested if 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"O\\": {
            \\"big\\": false,
            \\"red\\": false
        },
        \\"N\\": [
            0
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            1
        ]
    }
}",
  "html": "
<!--0-->
<!--0-->
",
}
`;
